sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBuffer.scala

```diff
@@ -72,6 +79,19 @@ trait RapidsBuffer extends AutoCloseable {
   /** The storage tier for this buffer */
   val storageTier: StorageTier
 
+  /**
+   * Called when a batch is spilled from one storage tier to another. This is intended to only
+   * be used for metrics gathering in parts of the GPU plan that can spill. No GPU memory
+   * should ever be allocated from this callback, blocking in this function is strongly
+   * discouraged. It should be as light weight as possible. It takes three arguments
+   * <ul>
+   * <li><code>from</code> the storage tier the data is being spilled from.</li>
+   * <li><code>to</code> the storage tier the data is being spilled to.</li>
+   * <li><code>amount</code> the amount of data in bytes that is spilled.</li>
+   * </ul>
+   */
+  val spillCallback: (StorageTier, StorageTier, Long) => Unit
```

Comment:
I don't default it here because the only time it needs to be defaulted is in the degenerate case and tests. I would rather be cautious and force someone adding a new buffer to decide if they need it or not, but with RapidsBufferBase it may be a moot point.